jans-cedarling/cedarling/src/log/log_level.rs

```diff
@@ -0,0 +1,104 @@
+/*
+ * This software is available under the Apache-2.0 license.
+ * See https://www.apache.org/licenses/LICENSE-2.0.txt for full text.
+ *
+ * Copyright (c) 2024, Gluu, Inc.
+ */
+
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt;
+use std::str::FromStr;
+
+/// Log levels
+/// Fatal level is the highest, trace is lowest
+#[derive(Debug, PartialEq, PartialOrd, Eq, Ord, Copy, Clone, Default)]
+pub enum LogLevel {
+    /// Fatal level
+    FATAL = 5,
+    /// Error level
+    ERROR = 4,
+    /// Warn level
+    #[default]
+    WARN = 3,
+    /// Info level
+    INFO = 2,
+    /// Debug level
+    DEBUG = 1,
+    /// Trace level
+    TRACE = 0,
+}
+
+impl LogLevel {
+    fn as_str(&self) -> &'static str {
+        match self {
+            LogLevel::FATAL => "FATAL",
+            LogLevel::ERROR => "ERROR",
+            LogLevel::WARN => "WARN",
+            LogLevel::INFO => "INFO",
+            LogLevel::DEBUG => "DEBUG",
+            LogLevel::TRACE => "TRACE",
+        }
+    }
+}
```

Comment:
We might be able just use `derive(derive_more::Display)` here instead of implementing an `as_str` function
```suggestion
/// Log levels
/// Fatal level is the highest, trace is lowest
#[derive(Debug, PartialEq, PartialOrd, Eq, Ord, Copy, Clone, Default, Display)]
pub enum LogLevel {
    /// Fatal level
    FATAL = 5,
    /// Error level
    ERROR = 4,
    /// Warn level
    #[default]
    WARN = 3,
    /// Info level
    INFO = 2,
    /// Debug level
    DEBUG = 1,
    /// Trace level
    TRACE = 0,
}
```

Reply:
Using `derive(derive_more::Display)` we need to use cast to `String` what is allocation on heap
```
impl Serialize for LogLevel {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(self.to_string().as_str())
    }
}
```
Do you think it is a good tradeoff ?

Reply:
oh, if you're using it like that and not printing it then having an `as_str` function might be better. Although, is that any different from doing `#[derive(Deserialize)]`?
```suggestion
#[derive(Debug, derive_more::Display, Default, Serialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum LogLevel {
    /// Fatal level
    Fatal = 5,
    /// Error level
    Error = 4,
    /// Warn level
    #[default]
    Warn = 3,
    /// Info level
    Info = 2,
    /// Debug level
    Debug = 1,
    /// Trace level
    Trace = 0,
}
```
I haven't looked into the `derive(Serialize)` implementation but it's less code if it's the same thing.

Reply:
I forgot about using macro serde::{Deserialize, Serialize}, it works OK with it.